packages/langium/test/syntax-tree.test.ts

```diff
@@ -0,0 +1,260 @@
+/******************************************************************************
+ * Copyright 2024 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { describe, test } from 'vitest';
+import type { AstNode, AstNodeTypesWithCrossReferences, AstTypeList, CrossReferencesOfAstNodeType, Reference } from '../src/syntax-tree.js';
+
+describe('Utility types revealing cross-reference properties', () => {
+
+    /**
+     * The tests listed below don't check correctness of 'CrossReferencesOfAstNodeType'
+     *  and 'AstNodeTypesWithCrossReferences' by executing code and comparing results.
+     * Instead, they check if the types are correctly inferred by TypeScript.
+     *
+     * Hence, test failure are indicated by TypeScript errors, while the absence of any compile errors denotes success.
+     *
+     * Note: a value of type 'never' can be assigned to any other type, but not vice versa.
+     * In order to make sure 'never' _is not_ derived where they shouldn't be, esp. for the 'props' variables, 'props.length' is used.
+     * In order to make sure 'never' _is_ derived where it should be, the function 'checkNever' defined below is used.
+     */
+
+    // below follow some type definitions as produced by the Langium generator
+
+    interface NoRefs extends AstNode {
+        $type: 'NoRefs';
+        name: string;
+    }
+
+    interface SingleRef extends AstNode {
+        $type: 'SingleRef';
+        singleRef: Reference<NoRefs>;
+    }
+
+    interface OptionalSingleRef extends AstNode {
+        $type: 'OptionalSingleRef';
+        optionalSingleRef?: Reference<NoRefs>;
+    }
+
+    interface MultiRef extends AstNode {
+        $type: 'MultiRef';
+        multiRef: Array<Reference<NoRefs>>;
+    }
+
+    interface OptionalMultiRef extends AstNode {
+        $type: 'OptionalMultiRef';
+        optionalMultiRef?: Array<Reference<NoRefs>>;
+    }
+
+    type PlainAstTypes = {
+        NoRefs: NoRefs,
+        SingleRef: SingleRef,
+        OptionalSingleRef: OptionalSingleRef,
+        MultiRef: MultiRef,
+        OptionalMultiRef: OptionalMultiRef
+    }
+
+    const getAnyInstanceOfType = <T extends AstTypeList<T>>()  => <AstNodeTypesWithCrossReferences<T>>{};
+    const getCrossRefProps =     <T extends AstNode>(_type: T) => <CrossReferencesOfAstNodeType<T>>'';
+    const checkNever =           (props: never)                => props;
+
+    test('Should not reveal cross-ref properties for NoRefs', () => {
+        const props = getCrossRefProps(<NoRefs>{});
+        checkNever(props);
```

Comment:
Suggestion: Instead using a custom `checkNever` function, you can use vitest's type testing utilities. See [their type testing guide](https://vitest.dev/guide/testing-types).
```suggestion
        expectTypeOf(props).toBeNever();
```

Reply:
Switched to that API, it's actually quite cool! 💡